mlflow/pyfunc/utils/data_validation.py

```diff
@@ -0,0 +1,98 @@
+import warnings
+from functools import lru_cache, wraps
+
+from mlflow.models.signature import _extract_type_hints, _is_context_in_predict_function_signature
+from mlflow.types.type_hints import (
+    _convert_data_to_type_hint,
+    _infer_schema_from_type_hint,
+    _signature_cannot_be_inferred_from_type_hint,
+    _validate_example_against_type_hint,
+)
+from mlflow.utils.annotations import filter_user_warnings_once
+
+
+def pyfunc(func):
+    """
+    A decorator that forces data validation against type hint of the input data
+    in the wrapped method. It is no-op if the type hint is not supported by MLflow.
+
+    .. note::
+        The function that applies this decorator must be a valid `predict` function
+        of `mlflow.pyfunc.PythonModel`, or a callable that takes a single input.
+    """
+
```

Comment:
What do you think about the idea of doing a quick validation here to ensure tke function has expected arguments (e.g. if it has model_input kwargs or single arg otherwise). This way users can know the error as soon sa defining the function or pyfunc class.

Reply:
Yep make sense

Reply:
Currently we have some warnings inside the wrapper https://github.com/mlflow/mlflow/blob/747ac6d3a55e8f9cacb66e579aa82489909fe9d7/mlflow/pyfunc/model.py#L645 I'll file a follow-up PR to move the warnings here instead of the wrapper, don't want to introduce too many changes in this PR :)

Reply:
Follow-up is fine, but let's make sure it is checked when **defining** the function, not invoking it🙂